src/core/main.c

```diff
@@ -1704,24 +1704,33 @@ static void initialize_clock(void) {
                  * time concepts will be treated as UTC that way.
                  */
                 (void) clock_reset_timewarp();
+}
+
+static void initialize_clock_epoch(bool allow_backwards) {
+        int r;
+
+        /* Optionally reset the clock to the epoch. If we're just starting, consider CLOCK_CHANGE_BACKWARD.
+         * If we're reexecuting, only jump forward. This way we'll update the epoch if were running with a
+         * bad clock and we're reexecuting into a newer systemd with an updated epoch timestamp. But we'll
+         * not jump back if we were running for a long long time and get reexecuted. */
 
         ClockChangeDirection change_dir;
-        r = clock_apply_epoch(&change_dir);
+        r = clock_apply_epoch(allow_backwards, &change_dir);
         if (r > 0 && change_dir == CLOCK_CHANGE_FORWARD)
-                log_info("System time before build time, advancing clock.");
+                log_info("System time before epoch, advancing clock.");
         else if (r > 0 && change_dir == CLOCK_CHANGE_BACKWARD)
-                log_info("System time is further ahead than %s after build time, resetting clock to build time.",
+                log_info("System time is further ahead than %s after epoch, resetting clock to epoch.",
                          FORMAT_TIMESPAN(CLOCK_VALID_RANGE_USEC_MAX, USEC_PER_DAY));
         else if (r < 0 && change_dir == CLOCK_CHANGE_FORWARD)
-                log_error_errno(r, "Current system time is before build time, but cannot correct: %m");
+                log_error_errno(r, "Current system time is before epoch, but cannot correct: %m");
         else if (r < 0 && change_dir == CLOCK_CHANGE_BACKWARD)
-                log_error_errno(r, "Current system time is further ahead %s after build time, but cannot correct: %m",
+                log_error_errno(r, "Current system time is further ahead %s after epoch, but cannot correct: %m",
```

Comment:
Hmm, if you touch this, maybe use `switch (change_dir)`? Also, please make the error msg and normal log msg consistent (`System time is further ahead _than_ %s` vs `system time is further ahead %s`

Reply:
This is C, not Rust or Python! We can't have conditionalized mathes, and without that a switch would be very awkward.